Feature/pie menu/master

SebastianMeinberger opened the pull request:
The 3D PieMenu from VIAProMa in a much more versatile version.
Still missing:
-Unit Tests
-Documentation
-A nice example scene demonstrating the capabilities of the PieMenu

BenediktHensen reviewed (commented):
Can this feature only be tested on a HTC Vive? Is it also possible to see it in action in the editor?

SebastianMeinberger commented:
All scripts are now in the the i5.Toolkit.MixedReality.PieMenu namespace and the example scene is now in the proper folder.

Regarding the UI Problems:
I unfortunatly wasn't able to reproduce the error. Does this problem happen everytime on your machine? However, I changed something at the editor script that may have caused the problem, but I can't say for sure.

Regarding the testing without a Vive:
Yes, that is indeed possible, with the [Input Simulation Service of the MRTK](https://docs.microsoft.com/en-us/windows/mixed-reality/mrtk-unity/features/input-simulation/input-simulation-service). I changed the MRTK Settings for the PieMenu example scene, so that controller simulation is enabled there and adjusted some things for our use case .
Now when starting the application through the editor, you can use the simulated controller like previously the simulated hand (pressing shift will spawn the right controller, etc.)
But the problem is, that this simulation only allows to use the menu, trigger and grip functionallity, not the touchpad. Unfortunately, the only way I manged to implement to circumvent this, results in a really cumbersome workflow.

Here are some step by step instructions, to use the simulaton service (it is important to keep in mind, that the simulated controller only recives input events, as long as you keep holding LShift):

1. Presss T to toggle the right controller (otherwise all scripts regarding the controller will be reseted, once LShift is released)
2. Frist press and hold LShift and then press and hold M without relasing LShift. Now the PieMenu should open
3. Move the controller with the mouse so you select the Menu entry you want
4. Release M. Now the icon of the MenuEntry should be displayed on the controller
5. Now you can use the trigger through mouse leftclick and the grip through pressing G
6. (Next steps are for using the touchpad) release shift
7. Without stopping the application: Click on the PieMenu manager in the scene hirachy
8. Go to the actions tab
9. Change the trigger input action to "None"
10. Change the touchpad press action to "Select". Should now look like this
![grafik](https://user-images.githubusercontent.com/59895994/116402088-471c5a80-a82c-11eb-9436-e0fd8efbc862.png)
11. (Still in the PieMenu Manager) Go to the debug tab
12. Press the button corresponding to the direction you want to press the touchpad
13. Focus the application again
14. When you now leftclick (as long as you hold LShift) it wont perform the select action anymore, but press the touchpad in the previosly definded direction. To change the direction, just press the corresponding button in the debug tab, without repeating the previous steps.

I know that this is really cumbersome and anoying, but I am afraid, this is the best method I found. The other methods are:
1. Triggering the actions through some other MRTK device, like an XBox Controller. But this would only shift the hardware availability problem to other hardware and would require to soften the safty mechanism that prevent input sources from picking up events from other input sources, wich could lead to problems in the future.
2. Simultaneously using default Unity Input Events, to react to keyboard inputs and not just the MRTK ones. Would require majaor changes and it wouldn't be a real simulation, but a circumvention, which could lead to wrong debug results.

BenediktHensen reviewed (changes requested):
Really well done! I can't wait to include this in the toolkit. This is really useful functionality and it is really well executed.

There are just a couple of minor things to address before merging, mainly typos and small optimizations. Please, check again that you only use XML documentation comments for public methods. If we generate documentation from them, we only want to show the publicly available methods to developers. For private methods, you can stick to plain C# comments which will not show up in the documentation but still help developers of the toolkit.
There is also a typo in the file Assets/i5 Toolkit for Mixed Reality/Runtime/PieMenu/Textures/Corsor.png to Cursor.png.

SebastianMeinberger commented on Assets/i5 Toolkit for Mixed Reality/Runtime/PieMenu/Scripts/PieMenuServiceBootstraper.cs:
```diff
@@ -0,0 +1,26 @@
+        protected override void UnRegisterServices()
+        {
+            ServiceManager.RemoveService<CommandStackService>();
+        }
```
Yes, I forgott to unregister that. Changed it.

SebastianMeinberger replied:
https://github.com/rwth-acis/i5-Toolkit-for-Mixed-Reality/pull/1/commits/5f8a26f23106ed400634b983413a7bbd84f8afca

SebastianMeinberger commented on Assets/i5 Toolkit for Mixed Reality/Editor/PieMenuEditor.cs:
```diff
@@ -0,0 +1,163 @@
+            }
+            if (GUILayout.Button("Menu Entries", state == PieMenuInsepectorState.MenuEntries ? buttonMarked : buttonNormal))
+            {
+                state = PieMenuInsepectorState.MenuEntries;
```
https://github.com/rwth-acis/i5-Toolkit-for-Mixed-Reality/pull/1/commits/c76258463250515b83bda2e53b59806dad1d2f52

SebastianMeinberger commented on Assets/i5 Toolkit for Mixed Reality/Runtime/PieMenu/Scripts/IObjectTransformer.cs:
```diff
@@ -0,0 +1,11 @@
+{
+    public interface IObjectTransformer
+    {
+        GameObject transformObject(GameObject objectToTransform, string toolName);
```
https://github.com/rwth-acis/i5-Toolkit-for-Mixed-Reality/pull/1/commits/77bbec5b50dc97f1844edef8473bc93871202b34

SebastianMeinberger commented on Assets/i5 Toolkit for Mixed Reality/Runtime/PieMenu/Scripts/PieMenuRenderer.cs:
```diff
@@ -0,0 +1,149 @@
+        /// Convert an index of the MenuEntry array to the corrosponding rotation on the PieMenu
+        /// </summary>
+        /// <param name="number"></param> The index from the MenuEntry array
+        /// <returns></returns> The corresponding rotation on the PieMenu
```
https://github.com/rwth-acis/i5-Toolkit-for-Mixed-Reality/pull/1/commits/416bf012555af301f12490fd5c8982c5e1b63fa6

SebastianMeinberger commented on Assets/i5 Toolkit for Mixed Reality/Runtime/PieMenu/Scripts/PieMenuRenderer.cs:
```diff
@@ -0,0 +1,149 @@
+        List<GameObject> pieMenuPieces;
+        public int currentlyHighlighted { private set; get; }
+
+        GameObject instiatedMenuCursor;
```
https://github.com/rwth-acis/i5-Toolkit-for-Mixed-Reality/pull/1/commits/3600ed8d000233e8ccac2c1c893bb60248d2e4a7

SebastianMeinberger commented on Assets/i5 Toolkit for Mixed Reality/Runtime/PieMenu/Scripts/Toolactions/ActionHelperFunctions.cs:
```diff
@@ -0,0 +1,125 @@
+            return pointer?.Controller?.Visualizer?.GameObjectProxy?.GetComponentInChildren<ViveWandVirtualTool>();
+        }
+
+        public static ViveWandVirtualTool GetVirtualTool(IMixedRealityInputSource inputSource)
```
https://github.com/rwth-acis/i5-Toolkit-for-Mixed-Reality/pull/1/commits/79c352aab25c91e7062a246d14b124636a8fc2ab

SebastianMeinberger commented on Assets/i5 Toolkit for Mixed Reality/Runtime/PieMenu/Scripts/Toolactions/IToolAction.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace i5.Toolkit.MixedReality.PieMenu
+{
+    public interface IToolAction
+    {
+        void DoAction();
+        void UndoAction();
```
https://github.com/rwth-acis/i5-Toolkit-for-Mixed-Reality/pull/1/commits/2dc4216afe54385caf839df55a8d6aa9e7387187

SebastianMeinberger commented on Assets/i5 Toolkit for Mixed Reality/Runtime/PieMenu/Scripts/ViveWandTeleporter.cs:
```diff
@@ -0,0 +1,65 @@
+
+            StopCoroutine("DisableDescriptions");
+            //Waits descriptionShowTime befor disabling the descriptions
+            StartCoroutine("DisableDescriptions");
```
https://github.com/rwth-acis/i5-Toolkit-for-Mixed-Reality/pull/1/commits/3fc1d0408cb96a0898ae5b84b90f7291fc35d305

SebastianMeinberger commented on Assets/i5 Toolkit for Mixed Reality/Runtime/PieMenu/Scripts/ViveWandTeleporter.cs:
```diff
@@ -0,0 +1,65 @@
+        /// </summary>
+        private void OnEnable()
+        {
+            StartCoroutine("SetOwnSource");
```
https://github.com/rwth-acis/i5-Toolkit-for-Mixed-Reality/pull/1/commits/3fc1d0408cb96a0898ae5b84b90f7291fc35d305

SebastianMeinberger commented on Assets/i5 Toolkit for Mixed Reality/Runtime/PieMenu/Scripts/ViveWandVirtualTool.cs:
```diff
@@ -0,0 +1,303 @@
+
+            StopCoroutine("DisableDescriptions");
+            //Waits descriptionShowTime befor disabling the descriptions
+            StartCoroutine("DisableDescriptions");
```
https://github.com/rwth-acis/i5-Toolkit-for-Mixed-Reality/pull/1/commits/3fc1d0408cb96a0898ae5b84b90f7291fc35d305

SebastianMeinberger commented on Assets/i5 Toolkit for Mixed Reality/Runtime/PieMenu/Scripts/ViveWandVirtualTool.cs:
```diff
@@ -0,0 +1,303 @@
+        /// </summary>
+        private void OnEnable()
+        {
+            StartCoroutine("SetOwnSource");
```
https://github.com/rwth-acis/i5-Toolkit-for-Mixed-Reality/pull/1/commits/3fc1d0408cb96a0898ae5b84b90f7291fc35d305

SebastianMeinberger commented on Assets/i5 Toolkit for Mixed Reality/Runtime/PieMenu/Scripts/ViveWandTeleporter.cs:
```diff
@@ -0,0 +1,65 @@
+        /// <summary>
+        /// Triggerd when an input action of type float changes its value. Used for the grip button.
+        /// </summary>
+        /// <param name="eventData"></param>
```
https://github.com/rwth-acis/i5-Toolkit-for-Mixed-Reality/pull/1/commits/a1f81b530b0f77ae3f6649b909b095597cde1ff3

SebastianMeinberger commented on Documentation/manual/PieMenu.md:
```diff
@@ -0,0 +1,117 @@
+
+
+## Creating a Menu Entry
+You can either follow these general instructions or start by building a small sample scene with the help of these TODO(link somehow) step by step instructions.
```
https://github.com/rwth-acis/i5-Toolkit-for-Mixed-Reality/pull/1/commits/9331c2678f75007e8aa6397c890483779bf52066
